DSharpPlus.Lavalink/LavalinkExtension.cs

```diff
@@ -71,15 +78,97 @@ public async Task<LavalinkNodeConnection> ConnectAsync(LavalinkConfiguration con
         }
 
         /// <summary>
-        /// Gets the lavalink node connection for specified endpoint.
+        /// Gets the Lavalink node connection for the specified endpoint.
         /// </summary>
         /// <param name="endpoint">Endpoint at which the node resides.</param>
         /// <returns>Lavalink node connection.</returns>
         public LavalinkNodeConnection GetNodeConnection(ConnectionEndpoint endpoint)
-            => this.ConnectedNodes.ContainsKey(endpoint) ? this.ConnectedNodes[endpoint] : null;
+            => this._connectedNodes.ContainsKey(endpoint) ? this._connectedNodes[endpoint] : null;
+
+        /// <summary>
+        /// Gets a Lavalink node connection based on load balancing and an optional voice region.
+        /// </summary>
+        /// <param name="region">The region to compare with the node's <see cref="LavalinkConfiguration.Region"/>, if any.</param>
+        /// <returns>The least load affected node connection, or null if no nodes are present.</returns>
+        public LavalinkNodeConnection GetIdealNodeConnection(DiscordVoiceRegion region = null)
+        {
+            if (this._connectedNodes.Count <= 1)
+                return this._connectedNodes.Values.FirstOrDefault();
+
+            var nodes = this._connectedNodes.Values.ToArray();
+
+            if (region != null)
+            {
+                var regionPredicate = new Func<LavalinkNodeConnection, bool>(x => x.Region == region);
+
+                if (nodes.Any(regionPredicate))
+                    nodes = nodes.Where(regionPredicate).ToArray();
+
+                if (nodes.Count() <= 1)
+                    return nodes.FirstOrDefault();
+            }
+
+            return this.FilterByLoad(nodes);
+        }
+
+        /// <summary>
+        /// Gets a Lavalink guild connection from a <see cref="DiscordGuild"/>.
+        /// </summary>
+        /// <param name="guild">The guild the connection is on.</param>
+        /// <returns>The found guild connection, or null if one could not be found.</returns>
+        public LavalinkGuildConnection GetGuildConnection(DiscordGuild guild)
+        {
+            var nodes = this._connectedNodes.Values;
+            var node = nodes.FirstOrDefault(x => x._connectedGuilds.ContainsKey(guild.Id));
+            return node?.GetGuildConnection(guild);
+        }
+
+        private LavalinkNodeConnection FilterByLoad(LavalinkNodeConnection[] nodes)
+        {
+            Array.Sort(nodes, (a, b) => 
```

Comment:
It seems like it might be a good idea to cache those.

+brace style on L145

Reply:
What should be cached?

Reply:
The penalties, with a TTL. Then again, if this doesn't happen too frequently, the caching might hurt it.

Reply:
If we cache this, wouldn't we have to update it every time the statistics received event is fired then?